File: DataStructureAndAlgorithm/DataStructures/StacksAndQueues/C++/Queue.cpp
#include "Queue.h"
#include <stdexcept>

void Queue::enqueue(int data) {
    deq.push_back(data);
}

int Queue::dequeue() {
    if (isEmpty()) {
        throw std::out_of_range("Queue is empty");
    }
    int val = deq.front();
    deq.pop_front();
    return val;
}

int Queue::front() {
    if (isEmpty()) {
        throw std::out_of_range("Queue is empty");
    }
    return deq.front();
}

bool Queue::isEmpty() {
    return deq.empty();
}

int Queue::size() {
    return deq.size();
}

File: DataStructureAndAlgorithm/DataStructures/StacksAndQueues/Python/queue.py
from collections import deque

class Queue:
    def __init__(self):
        self.items = deque()

    def is_empty(self):
        return not self.items

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.popleft()
        raise IndexError("dequeue from empty queue")

    def front(self):
        if not self.is_empty():
            return self.items[0]
        raise IndexError("front from empty queue")

    def size(self):
        return len(self.items)
